sw/device/tests/dif/dif_hmac_sanitytest.c

```diff
@@ -0,0 +1,203 @@
+  const char *dp = data;
+  size_t sent_bytes;
+
+  while (dp - data < len) {
+    dif_hmac_fifo_result_t res =
+        dif_hmac_fifo_push(hmac, dp, len - (dp - data), &sent_bytes);
+
+    CHECK(res != kDifHmacFifoBadArg,
+          "Invalid arguments encountered while pushing to FIFO.");
+
+    if (res == kDifHmacFifoFull) {
+      ++fifo_fill_count;
+    } else {
+      CHECK(res != kDifHmacFifoBadArg,
+            "Invalid arguments encountered while pushing to FIFO.");
+      CHECK(res == kDifHmacFifoOk,
+            "Unknown error encountered while pushing to FIFO.");
+    }
+
+    CHECK(fifo_fill_count <= MAX_FIFO_FILL,
+          "FIFO filled up too may times, giving up.");
+
+    dp += sent_bytes;
+  }
+}
+
+/** Spin while the HMAC FIFO still has entries in it. Once the FIFO is empty we
+ * can check the message length. Return `true` if there are no errors.
+ */
+static void wait_for_fifo_empty(const dif_hmac_t *hmac) {
+  uint32_t fifo_depth;
+  do {
+    dif_hmac_result_t res = dif_hmac_fifo_count_entries(hmac, &fifo_depth);
+
+    CHECK(res != kDifHmacBadArg,
+          "Invalid arguments encountered checking FIFO depth.");
+    CHECK(res == kDifHmacOk, "Unknown error encountered checking FIFO depth.");
+  } while (fifo_depth > 0);
+}
+
+/**
+ * Read and compare the length of the message in the HMAC engine to the length
+ * of the message sent in bits.
+ */
+static void check_message_length(const dif_hmac_t *hmac,
+                                 uint64_t expected_sent_bits) {
+  uint64_t sent_bits;
+  dif_hmac_result_t res = dif_hmac_get_message_length(hmac, &sent_bits);
+
+  CHECK(res != kDifHmacBadArg,
+        "Invalid arguments encountered checking message length.");
+  CHECK(res == kDifHmacOk,
+        "Unknown error encountered checking message length.");
+
+  // TODO: Support 64-bit integers in logging.
+  CHECK(expected_sent_bits == sent_bits,
+        "Message length mismatch. Expected %u bits but got %u bits.",
+        (uint32_t)expected_sent_bits, (uint32_t)sent_bits);
+}
+
+/**
+ * Kick off the HMAC (or SHA256) run.
+ */
+static void run_hmac(const dif_hmac_t *hmac) {
+  dif_hmac_result_t res = dif_hmac_process(hmac);
+
+  CHECK(res != kDifHmacBadArg, "Invalid arguments encountered running HMAC.");
+  CHECK(res == kDifHmacOk, "Unknown error encountered running HMAC.");
+}
+
+/**
+ * Read the HMAC digest and compare it to the expected result.
+ */
+static void check_digest(const dif_hmac_t *hmac,
+                         const uint32_t *expected_digest) {
+  dif_hmac_digest_t digest_result;
+  bool hmac_done = false;
+  do {
+    dif_hmac_digest_result_t res = dif_hmac_digest_read(hmac, &digest_result);
+
+    CHECK(res != kDifHmacDigestBadArg,
+          "Invalid arguments encountered reading HMAC digest.");
+
+    hmac_done = (res != kDifHmacDigestProcessing);
+
+    if (hmac_done) {
+      CHECK(res == kDifHmacDigestOk,
+            "Unknown error encountered reading HMAC digest.");
+    }
+  } while (!hmac_done);
```

Comment:
AFAICT this can just be written as a while() loop, which is less confusing than a do/while. You can punt this to a followup.